File: BFS.py
# Complete program for Breadth First Search Traversal (BFS).

from collections import deque

def bfs(graph, start_vertex):
    visited = set()
    queue = deque()

    visited.add(start_vertex)
    queue.append(start_vertex)

    print("Breadth First Traversal starting from vertex", start_vertex, ":")

    while queue:
        vertex = queue.popleft()
        print(vertex, end=" ")

        for neighbor in graph[vertex]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)
    print()

def main():
    graph = {}
    vertices = int(input("Enter number of vertices: "))

    for i in range(vertices):
        edges = list(map(int, input(f"Enter neighbors of vertex {i} (space-separated): ").split()))
        graph[i] = edges

    start_vertex = int(input(f"Enter the starting vertex (0-{vertices-1}): "))
    bfs(graph, start_vertex)

if __name__ == "__main__":
    main()

File: BFS.c
// Complete program for Breadth First Search Traversal (BFS).

#include <stdio.h>
#include <stdlib.h>

#define MAX 100

// Queue structure for BFS
struct Queue {
    int items[MAX];
    int front;
    int rear;
};

struct Queue* createQueue() {
    struct Queue* q = (struct Queue*)malloc(sizeof(struct Queue));
    q->front = -1;
    q->rear = -1;
    return q;
}

int isEmpty(struct Queue* q) {
    if (q->rear == -1)
        return 1;
    else
        return 0;
}

void enqueue(struct Queue* q, int value) {
    if (q->rear == MAX - 1)
        printf("\nQueue is Full!!");
    else {
        if (q->front == -1)
            q->front = 0;
        q->rear++;
        q->items[q->rear] = value;
    }
}

int dequeue(struct Queue* q) {
    int item;
    if (isEmpty(q)) {
        printf("Queue is empty");
        item = -1;
    } else {
        item = q->items[q->front];
        q->front++;
        if (q->front > q->rear) {
            q->front = q->rear = -1;
        }
    }
    return item;
}

// BFS function
void bfs(int adj[][MAX], int vertices, int startVertex) {
    int visited[MAX] = {0};
    struct Queue* q = createQueue();

    visited[startVertex] = 1;
    enqueue(q, startVertex);

    printf("Breadth First Traversal starting from vertex %d:\n", startVertex);

    while (!isEmpty(q)) {
        int currentVertex = dequeue(q);
        printf("%d ", currentVertex);

        for (int i = 0; i < vertices; i++) {
            if (adj[currentVertex][i] == 1 && !visited[i]) {
                visited[i] = 1;
                enqueue(q, i);
            }
        }
    }
    printf("\n");
}

int main() {
    int vertices;
    int adj[MAX][MAX];
    int startVertex;

    printf("Enter number of vertices: ");
    scanf("%d", &vertices);

    printf("Enter adjacency matrix:\n");
    for (int i = 0; i < vertices; i++) {
        for (int j = 0; j < vertices; j++) {
            scanf("%d", &adj[i][j]);
        }
    }

    printf("Enter the starting vertex (0-%d): ", vertices-1);
    scanf("%d", &startVertex);

    bfs(adj, vertices, startVertex);

    return 0;
}
